pt/lessons/specifics/debugging.md

```diff
@@ -1,23 +1,162 @@
 
-Às vezes, os bugs estão presentes em nosso projeto, é por isso que precisamos investigar o problema. Nesta sessão vamos aprender sobre como fazer debug no código Elixir, bem como ferramentas de análise de estática para ajudar a encontrar possíveis bugs.
+Bugs são problemas comuns em qualquer projeto, é por isso que precisamos da depuração.
+
+Nesta sessão vamos aprender sobre como fazer depuração no código Elixir, bem como ferramentas de análises estáticas para ajudar a encontrar possíveis bugs.
 
 {% include toc.html %}
 
-# Dialyxir e Dialyzer
+## IEx
+
+A ferramenta mais direta que nós temos para depurar código Elixir é o IEx
+
+Mas não se deixa enganar por sua simplicidade - você pode resolver a maioria dos problemas da sua aplicação com ele.
+
+IEx significa `Elixir's Interactive Shell`.
+
+Você pode já ter visto o IEx em uma das lições anteriores como [Básicos](../../basisc/basics) onde nós executamos código interativo no shell.
+
+A ideia aqui é simples.
+
+Você inicia o shell interativo no contexto do local que deseja depurar.
+
+Vamos tentar.
+
+```elixir
+defmodule TestMod do
+  def sum([a, b]) do
+    b = 0
+
+    a + b
+  end
+end
+
+IO.puts(TestMod.sum([34, 65]))
+```
+
+E se você executar isso - você receberá uma saída aparente de '34':
+
+```shell
+$ elixir test.exs
+warning: variable "b" is unused (if the variable is not meant to be used, prefix it with an underscore)
+ test.exs:2
+
+34
+```
+
+Mas agora vamos para a parte empolgante - a depuração.
+
+Insira `require IEx: Iex.pry` depois da linha com `b = 0` e vamos tentar executar o código novamente.
+
+Você vai receber algo assim:
+
+```shell
+$ elixir test.exs
+warning: variable "b" is unused (if the variable is not meant to be used, prefix it with an underscore)
+  test.exs:2
+
+Cannot pry #PID<0.92.0> at TestMod.sum/1 (test.exs:5). Is an IEx shell running?
+34
+```
+
+Você deve perceber aquela mensagem crucial.
+
+Quando estiver executando uma aplicação, como sempre, o IEx gera essa mensagem em vez de bloquear a execução do programa.
+
+Para executar ela apropriadamente você precisa prefixar seu comando com `iex -S`.
+
+O que isso faz é executar `mix` dentro do comando `iex`, fazendo com que a aplicação seja executada em um modo especial, de tal jeito que chame `IEx.pry` para parar a execução da aplicação.
+
+Por exemplo, use `iex -S mix phx.serve` para depurar sua aplicação Phoenix. No nosso caso será `iex -r test.exs` para requerir o arquivo:
+
+```shell
+$ iex -r test.exs
+Erlang/OTP 21 [erts-10.3.1] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1] [hipe] [dtrace]
+
+warning: variable "b" is unused (if the variable is not meant to be used, prefix it with an underscore)
+  test.exs:2
+
+Request to pry #PID<0.107.0> at TestMod.sum/1 (test.exs:5)
+
+    3:     b = 0
+    4:
+    5:     require IEx; IEx.pry
+    6:
+    7:     a + b
+
+Allow? [Yn]
+```
+
+Depois de responder a confirmação com `y` ou pressionar Enter, você vai entrar no modo interativo.
+
+```shell
```

Comment:
```suggestion
```
```

In the original we don't use the "lang" of the block here. IMO we shouldn't use shell here as well, because the block is not exactly a shell as it includes the interactions of a iex session